packages/core/src/sync/index.ts

```diff
@@ -936,243 +817,442 @@ export const createSync = async (args: CreateSyncParameters): Promise<Sync> => {
-
-  const [remoteChainId, startBlock, latestBlock] = await Promise.all([
-    requestQueue.request({ method: "eth_chainId" }),
-    _eth_getBlockByNumber(requestQueue, { blockNumber: start }),
-    _eth_getBlockByNumber(requestQueue, { blockTag: "latest" }),
-  ]);
-
-  const endBlock =
-    end === undefined
-      ? undefined
-      : end > hexToBigInt(latestBlock.number)
-        ? ({
-            number: toHex(end),
-            hash: "0x",
-            parentHash: "0x",
-            timestamp: toHex(maxCheckpoint.blockTimestamp),
-          } as LightBlock)
-        : await _eth_getBlockByNumber(requestQueue, { blockNumber: end });
+  let unfinalizedBlocks: Omit<
+    Extract<RealtimeSyncEvent, { type: "block" }>,
+    "type"
+  >[] = [];
 
-  // Warn if the config has a different chainId than the remote.
-  if (hexToNumber(remoteChainId) !== network.chainId) {
-    common.logger.warn({
-      service: "sync",
-      msg: `Remote chain ID (${remoteChainId}) does not match configured chain ID (${network.chainId}) for network "${network.name}"`,
-    });
-  }
+  return async (event: RealtimeSyncEvent): Promise<RealtimeSyncEvent> => {
+    switch (event.type) {
+      case "block": {
+        syncProgress.current = event.block;
 
-  const finalizedBlockNumber = Math.max(
-    0,
-    hexToNumber(latestBlock.number) - network.finalityBlockCount,
-  );
+        common.logger.debug({
+          service: "sync",
+          msg: `Updated '${network.name}' current block to ${hexToNumber(event.block.number)}`,
+        });
 
-  const finalizedBlock = await _eth_getBlockByNumber(requestQueue, {
-    blockNumber: finalizedBlockNumber,
-  });
+        common.metrics.ponder_sync_block.set(
+          { network: network.name },
+          hexToNumber(syncProgress.current.number),
+        );
 
-  return {
-    start: startBlock,
-    end: endBlock,
-    finalized: finalizedBlock,
+        unfinalizedBlocks.push(event);
+
+        return event;
+      }
+
+      case "finalize": {
+        // Newly finalized range
+        const finalizedInterval = [
+          hexToNumber(syncProgress.finalized.number),
+          hexToNumber(event.block.number),
+        ] satisfies Interval;
+
+        syncProgress.finalized = event.block;
+
+        common.logger.debug({
+          service: "sync",
+          msg: `Updated '${network.name}' finalized block to ${hexToNumber(event.block.number)}`,
+        });
+
+        // Remove all finalized data
+
+        const finalizedBlocks = unfinalizedBlocks.filter(
+          ({ block }) =>
+            hexToNumber(block.number) <= hexToNumber(event.block.number),
+        );
+
+        unfinalizedBlocks = unfinalizedBlocks.filter(
+          ({ block }) =>
+            hexToNumber(block.number) > hexToNumber(event.block.number),
+        );
+
+        // Add finalized blocks, logs, transactions, receipts, and traces to the sync-store.
+
+        await Promise.all([
```

Comment:
This ought to use a database transaction (including the interval insert statement).

Can tackle alongside the sync store improvements.

Reply:
Agreed